File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/process/traversal-strategy.ts
/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */

/**
 * @author Jorge Bay Gondra
 */

import { RemoteConnection } from '../driver/remote-connection.js';
import { Traversal } from './traversal.js';

export class TraversalStrategies {
  readonly strategies: TraversalStrategy[];

  /**
   * Creates a new instance of TraversalStrategies.
   * @param {TraversalStrategies} [parent] The parent strategies from where to clone the values from.
   * @constructor
   */
  constructor(parent?: TraversalStrategies) {
    if (parent) {
      // Clone the strategies
      this.strategies = [...parent.strategies];
    } else {
      this.strategies = [];
    }
  }

  /** @param {TraversalStrategy} strategy */
  addStrategy(strategy: TraversalStrategy) {
    this.strategies.push(strategy);
  }

  /** @param {TraversalStrategy} strategy */
  removeStrategy(strategy: TraversalStrategy) {
    const idx = this.strategies.findIndex((s) => s.fqcn === strategy.fqcn);
    if (idx !== -1) {
      return this.strategies.splice(idx, 1)[0];
    }

    return undefined;
  }

  /**
   * @param {Traversal} traversal
   * @returns {Promise}
   */
  applyStrategies(traversal: Traversal) {
    // Apply all strategies serially
    return this.strategies.reduce(
      (promise, strategy) => promise.then(() => strategy.apply(traversal)),
      Promise.resolve(),
    );
  }
}

export type TraversalStrategyConfiguration = any;

export abstract class TraversalStrategy {
  connection?: RemoteConnection;

  /**
   * @param {String} fqcn fully qualified class name in Java of the strategy
   * @param {TraversalStrategyConfiguration} configuration for the strategy
   */
  constructor(
    public fqcn: string,
    public configuration: TraversalStrategyConfiguration = {},
  ) {}

  /**
   * @abstract
   * @param {Traversal} traversal
   * @returns {Promise}
   */
  async apply(traversal: Traversal): Promise<void> {}
}

export class ConnectiveStrategy extends TraversalStrategy {
  constructor() {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.decoration.ConnectiveStrategy');
  }
}

export class ElementIdStrategy extends TraversalStrategy {
  constructor() {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.decoration.ElementIdStrategy');
  }
}

export class HaltedTraverserStrategy extends TraversalStrategy {
  /**
   * @param {String} haltedTraverserFactory full qualified class name in Java of a {@code HaltedTraverserFactory} implementation
   */
  constructor(haltedTraverserFactory: string) {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.decoration.HaltedTraverserStrategy');
    if (haltedTraverserFactory !== undefined) {
      this.configuration['haltedTraverserFactory'] = haltedTraverserFactory;
    }
  }
}

export class OptionsStrategy extends TraversalStrategy {
  constructor(options: TraversalStrategyConfiguration) {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.decoration.OptionsStrategy', options);
  }
}

export class PartitionStrategy extends TraversalStrategy {
  /**
   * @param {Object} [options]
   * @param {String} [options.partitionKey] name of the property key to partition by
   * @param {String} [options.writePartition] the value of the currently write partition
   * @param {Array<String>} [options.readPartitions] list of strings representing the partitions to include for reads
   * @param {boolean} [options.includeMetaProperties] determines if meta-properties should be included in partitioning defaulting to false
   */
  constructor(options: TraversalStrategyConfiguration) {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.decoration.PartitionStrategy', options);
  }
}

export class SubgraphStrategy extends TraversalStrategy {
  /**
   * @param {Object} [options]
   * @param {GraphTraversal} [options.vertices] name of the property key to partition by
   * @param {GraphTraversal} [options.edges] the value of the currently write partition
   * @param {GraphTraversal} [options.vertexProperties] list of strings representing the partitions to include for reads
   * @param {boolean} [options.checkAdjacentVertices] enables the strategy to apply the {@code vertices} filter to the adjacent vertices of an edge.
   */
  constructor(options: TraversalStrategyConfiguration) {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.decoration.SubgraphStrategy', options);
    if (this.configuration.vertices instanceof Traversal) {
      this.configuration.vertices = this.configuration.vertices.bytecode;
    }
    if (this.configuration.edges instanceof Traversal) {
      this.configuration.edges = this.configuration.edges.bytecode;
    }
    if (this.configuration.vertexProperties instanceof Traversal) {
      this.configuration.vertexProperties = this.configuration.vertexProperties.bytecode;
    }
  }
}

export class ProductiveByStrategy extends TraversalStrategy {
  /**
   * @param {Object} [options]
   * @param {Array<String>} [options.productiveKeys] set of keys that will always be productive
   */
  constructor(options: TraversalStrategyConfiguration) {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.optimization.ProductiveByStrategy', options);
  }
}

export class VertexProgramStrategy extends TraversalStrategy {
  constructor(options: TraversalStrategyConfiguration) {
    super('org.apache.tinkerpop.gremlin.process.computer.traversal.strategy.decoration.VertexProgramStrategy', options);
  }
}

export class MatchAlgorithmStrategy extends TraversalStrategy {
  /**
   * @param matchAlgorithm
   */
  constructor(matchAlgorithm: string) {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.finalization.MatchAlgorithmStrategy');
    if (matchAlgorithm !== undefined) {
      this.configuration['matchAlgorithm'] = matchAlgorithm;
    }
  }
}

export class AdjacentToIncidentStrategy extends TraversalStrategy {
  constructor() {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.optimization.AdjacentToIncidentStrategy');
  }
}

export class FilterRankingStrategy extends TraversalStrategy {
  constructor() {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.optimization.FilterRankingStrategy');
  }
}

export class IdentityRemovalStrategy extends TraversalStrategy {
  constructor() {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.optimization.IdentityRemovalStrategy');
  }
}

export class IncidentToAdjacentStrategy extends TraversalStrategy {
  constructor() {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.optimization.IncidentToAdjacentStrategy');
  }
}

export class InlineFilterStrategy extends TraversalStrategy {
  constructor() {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.optimization.InlineFilterStrategy');
  }
}

export class LazyBarrierStrategy extends TraversalStrategy {
  constructor() {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.optimization.LazyBarrierStrategy');
  }
}

export class MatchPredicateStrategy extends TraversalStrategy {
  constructor() {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.optimization.MatchPredicateStrategy');
  }
}

export class OrderLimitStrategy extends TraversalStrategy {
  constructor() {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.optimization.OrderLimitStrategy');
  }
}

export class PathProcessorStrategy extends TraversalStrategy {
  constructor() {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.optimization.PathProcessorStrategy');
  }
}

export class PathRetractionStrategy extends TraversalStrategy {
  constructor() {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.optimization.PathRetractionStrategy');
  }
}

export class CountStrategy extends TraversalStrategy {
  constructor() {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.optimization.CountStrategy');
  }
}

export class RepeatUnrollStrategy extends TraversalStrategy {
  constructor() {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.optimization.RepeatUnrollStrategy');
  }
}

export class GraphFilterStrategy extends TraversalStrategy {
  constructor() {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.optimization.GraphFilterStrategy');
  }
}

export class EarlyLimitStrategy extends TraversalStrategy {
  constructor() {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.optimization.EarlyLimitStrategy');
  }
}

export class LambdaRestrictionStrategy extends TraversalStrategy {
  constructor() {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.verification.LambdaRestrictionStrategy');
  }
}

export class ReadOnlyStrategy extends TraversalStrategy {
  constructor() {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.verification.ReadOnlyStrategy');
  }
}

export class EdgeLabelVerificationStrategy extends TraversalStrategy {
  /**
   * @param {boolean} logWarnings determines if warnings should be written to the logger when verification fails
   * @param {boolean} throwException determines if exceptions should be thrown when verifications fails
   */
  constructor(logWarnings = false, throwException = false) {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.verification.EdgeLabelVerificationStrategy', {
      logWarnings: logWarnings,
      throwException: throwException,
    });
  }
}

export class ReservedKeysVerificationStrategy extends TraversalStrategy {
  /**
   * @param {boolean} logWarnings determines if warnings should be written to the logger when verification fails
   * @param {boolean} throwException determines if exceptions should be thrown when verifications fails
   * @param {Array<String>} keys the list of reserved keys to verify
   */
  constructor(logWarnings = false, throwException = false, keys = ['id', 'label']) {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.verification.ReservedKeysVerificationStrategy', {
      logWarnings: logWarnings,
      throwException: throwException,
      keys: keys,
    });
  }
}

export type SeedStrategyOptions = { seed: number };

export class SeedStrategy extends TraversalStrategy {
  /**
   * @param {SeedStrategyOptions} [options]
   * @param {number} [options.seed] the seed to provide to the random number generator for the traversal
   */
  constructor(options: SeedStrategyOptions) {
    super('org.apache.tinkerpop.gremlin.process.traversal.strategy.decoration.SeedStrategy', {
      seed: options.seed,
    });
  }
}

File: tinkubator/gremlin-dart/lib/process/traversal_strategy.dart
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements. See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership. The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License. You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied. See the License for the
// specific language governing permissions and limitations
// under the License.

import '../driver/transaction.dart';
import 'gremlin_lang.dart';
import 'traversal.dart';

class TraversalStrategies {
  final List<TraversalStrategy> strategies;

  TraversalStrategies([TraversalStrategies? parent])
      : strategies = List.of(parent?.strategies ?? []);

  void addStrategy(TraversalStrategy strategy) => strategies.add(strategy);

  void removeStrategy(TraversalStrategy strategy) =>
      strategies.removeWhere((s) => s.strategyName == strategy.strategyName);

  Future<void> applyStrategies(Traversal traversal) async {
    for (final s in strategies) {
      await s.apply(traversal);
    }
  }
}

abstract class TraversalStrategy {
  final String strategyName;
  final Map<String, dynamic> configuration;

  TraversalStrategy({String? strategyName, Map<String, dynamic>? configuration})
      : strategyName = strategyName ?? '',
        configuration = configuration ?? {};

  Future<void> apply(Traversal traversal) async {}
}

/// Resolves a strategy instance, class literal, or explicit name to the
/// server-side strategy name without relying on [Type.toString()].
String strategyNameOf(dynamic strategy) {
  if (strategy is String) return strategy;
  if (strategy is TraversalStrategy) return strategy.strategyName;

  const names = <Type, String>{
    RemoteStrategy: 'RemoteStrategy',
    OptionsStrategy: 'OptionsStrategy',
    PartitionStrategy: 'PartitionStrategy',
    SubgraphStrategy: 'SubgraphStrategy',
    SeedStrategy: 'SeedStrategy',
    ReadOnlyStrategy: 'ReadOnlyStrategy',
    VertexProgramStrategy: 'VertexProgramStrategy',
    AdjacentToIncidentStrategy: 'AdjacentToIncidentStrategy',
    ByModulatorOptimizationStrategy: 'ByModulatorOptimizationStrategy',
    ComputerFinalizationStrategy: 'ComputerFinalizationStrategy',
    ComputerVerificationStrategy: 'ComputerVerificationStrategy',
    ConnectiveStrategy: 'ConnectiveStrategy',
    CountStrategy: 'CountStrategy',
    EarlyLimitStrategy: 'EarlyLimitStrategy',
    ElementIdStrategy: 'ElementIdStrategy',
    FilterRankingStrategy: 'FilterRankingStrategy',
    GraphFilterStrategy: 'GraphFilterStrategy',
    IdentityRemovalStrategy: 'IdentityRemovalStrategy',
    IncidentToAdjacentStrategy: 'IncidentToAdjacentStrategy',
    InlineFilterStrategy: 'InlineFilterStrategy',
    LambdaRestrictionStrategy: 'LambdaRestrictionStrategy',
    LazyBarrierStrategy: 'LazyBarrierStrategy',
    MatchPredicateStrategy: 'MatchPredicateStrategy',
    MessagePassingReductionStrategy: 'MessagePassingReductionStrategy',
    OrderLimitStrategy: 'OrderLimitStrategy',
    PathProcessorStrategy: 'PathProcessorStrategy',
    PathRetractionStrategy: 'PathRetractionStrategy',
    ProductiveByStrategy: 'ProductiveByStrategy',
    ProfileStrategy: 'ProfileStrategy',
    ReferenceElementStrategy: 'ReferenceElementStrategy',
    RepeatUnrollStrategy: 'RepeatUnrollStrategy',
    StandardVerificationStrategy: 'StandardVerificationStrategy',
    VertexProgramRestrictionStrategy: 'VertexProgramRestrictionStrategy',
    ReservedKeysVerificationStrategy: 'ReservedKeysVerificationStrategy',
    EdgeLabelVerificationStrategy: 'EdgeLabelVerificationStrategy',
    MatchAlgorithmStrategy: 'MatchAlgorithmStrategy',
    HaltedTraverserStrategy: 'HaltedTraverserStrategy',
  };
  final name = names[strategy];
  if (name == null) {
    throw ArgumentError.value(strategy, 'strategy', 'Unsupported strategy');
  }
  return name;
}

// ---------------------------------------------------------------------------
// RemoteStrategy — wires a RemoteConnection to a traversal
// ---------------------------------------------------------------------------

class RemoteStrategy extends TraversalStrategy {
  final RemoteConnectionBase connection;

  RemoteStrategy(this.connection) : super(strategyName: 'RemoteStrategy');

  @override
  Future<void> apply(Traversal traversal) async {
    if (traversal.resultsStream != null) return;
    final remote = await connection.submit(traversal.getGremlinLang());
    traversal.resultsStream = remote.resultsStream;
  }
}

// Forward declaration to avoid circular import — concrete class lives in
// driver/remote_connection.dart.
abstract class RemoteConnectionBase {
  Future<RemoteTraversal> submit(GremlinLang gremlinLang);
}

abstract class TransactionCapableRemoteConnectionBase
    implements RemoteConnectionBase {
  Transaction tx([String? traversalSource]);
}

// ---------------------------------------------------------------------------
// Common strategies
// ---------------------------------------------------------------------------

class OptionsStrategy extends TraversalStrategy {
  OptionsStrategy([Map<String, dynamic>? options])
      : super(strategyName: 'OptionsStrategy', configuration: options);
}

class PartitionStrategy extends TraversalStrategy {
  PartitionStrategy({
    String? partitionKey,
    String? writePartition,
    List<String>? readPartitions,
    bool? includeMetaProperties,
  }) : super(strategyName: 'PartitionStrategy', configuration: {
          if (partitionKey != null) 'partitionKey': partitionKey,
          if (writePartition != null) 'writePartition': writePartition,
          if (readPartitions != null) 'readPartitions': readPartitions,
          if (includeMetaProperties != null)
            'includeMetaProperties': includeMetaProperties,
        });
}

class SubgraphStrategy extends TraversalStrategy {
  SubgraphStrategy({
    dynamic vertices,
    dynamic edges,
    dynamic vertexProperties,
    bool? checkAdjacentVertices,
  }) : super(strategyName: 'SubgraphStrategy', configuration: {
          if (vertices != null) 'vertices': vertices,
          if (edges != null) 'edges': edges,
          if (vertexProperties != null) 'vertexProperties': vertexProperties,
          if (checkAdjacentVertices != null)
            'checkAdjacentVertices': checkAdjacentVertices,
        });
}

class SeedStrategy extends TraversalStrategy {
  SeedStrategy({required dynamic seed})
      : super(
            strategyName: 'SeedStrategy',
            configuration: {'seed': seed is GInt ? seed.value : seed});
}

class ReadOnlyStrategy extends TraversalStrategy {
  ReadOnlyStrategy() : super(strategyName: 'ReadOnlyStrategy');
}

class VertexProgramStrategy extends TraversalStrategy {
  VertexProgramStrategy([Map<String, dynamic>? options])
      : super(strategyName: 'VertexProgramStrategy', configuration: options);
}

// Strategies without configuration.
class AdjacentToIncidentStrategy extends TraversalStrategy {
  AdjacentToIncidentStrategy()
      : super(strategyName: 'AdjacentToIncidentStrategy');
}

class ByModulatorOptimizationStrategy extends TraversalStrategy {
  ByModulatorOptimizationStrategy()
      : super(strategyName: 'ByModulatorOptimizationStrategy');
}

class ComputerFinalizationStrategy extends TraversalStrategy {
  ComputerFinalizationStrategy()
      : super(strategyName: 'ComputerFinalizationStrategy');
}

class ComputerVerificationStrategy extends TraversalStrategy {
  ComputerVerificationStrategy()
      : super(strategyName: 'ComputerVerificationStrategy');
}

class ConnectiveStrategy extends TraversalStrategy {
  ConnectiveStrategy() : super(strategyName: 'ConnectiveStrategy');
}

class CountStrategy extends TraversalStrategy {
  CountStrategy() : super(strategyName: 'CountStrategy');
}

class EarlyLimitStrategy extends TraversalStrategy {
  EarlyLimitStrategy() : super(strategyName: 'EarlyLimitStrategy');
}

class ElementIdStrategy extends TraversalStrategy {
  ElementIdStrategy() : super(strategyName: 'ElementIdStrategy');
}

class FilterRankingStrategy extends TraversalStrategy {
  FilterRankingStrategy() : super(strategyName: 'FilterRankingStrategy');
}

class GraphFilterStrategy extends TraversalStrategy {
  GraphFilterStrategy() : super(strategyName: 'GraphFilterStrategy');
}

class IdentityRemovalStrategy extends TraversalStrategy {
  IdentityRemovalStrategy() : super(strategyName: 'IdentityRemovalStrategy');
}

class IncidentToAdjacentStrategy extends TraversalStrategy {
  IncidentToAdjacentStrategy()
      : super(strategyName: 'IncidentToAdjacentStrategy');
}

class InlineFilterStrategy extends TraversalStrategy {
  InlineFilterStrategy() : super(strategyName: 'InlineFilterStrategy');
}

class LambdaRestrictionStrategy extends TraversalStrategy {
  LambdaRestrictionStrategy()
      : super(strategyName: 'LambdaRestrictionStrategy');
}

class LazyBarrierStrategy extends TraversalStrategy {
  LazyBarrierStrategy() : super(strategyName: 'LazyBarrierStrategy');
}

class MatchPredicateStrategy extends TraversalStrategy {
  MatchPredicateStrategy() : super(strategyName: 'MatchPredicateStrategy');
}

class MessagePassingReductionStrategy extends TraversalStrategy {
  MessagePassingReductionStrategy()
      : super(strategyName: 'MessagePassingReductionStrategy');
}

class OrderLimitStrategy extends TraversalStrategy {
  OrderLimitStrategy() : super(strategyName: 'OrderLimitStrategy');
}

class PathProcessorStrategy extends TraversalStrategy {
  PathProcessorStrategy() : super(strategyName: 'PathProcessorStrategy');
}

class PathRetractionStrategy extends TraversalStrategy {
  PathRetractionStrategy() : super(strategyName: 'PathRetractionStrategy');
}

class ProductiveByStrategy extends TraversalStrategy {
  ProductiveByStrategy() : super(strategyName: 'ProductiveByStrategy');
}

class ProfileStrategy extends TraversalStrategy {
  ProfileStrategy() : super(strategyName: 'ProfileStrategy');
}

class ReferenceElementStrategy extends TraversalStrategy {
  ReferenceElementStrategy() : super(strategyName: 'ReferenceElementStrategy');
}

class RepeatUnrollStrategy extends TraversalStrategy {
  RepeatUnrollStrategy() : super(strategyName: 'RepeatUnrollStrategy');
}

class StandardVerificationStrategy extends TraversalStrategy {
  StandardVerificationStrategy()
      : super(strategyName: 'StandardVerificationStrategy');
}

class VertexProgramRestrictionStrategy extends TraversalStrategy {
  VertexProgramRestrictionStrategy()
      : super(strategyName: 'VertexProgramRestrictionStrategy');
}

// Strategies with configuration.
class ReservedKeysVerificationStrategy extends TraversalStrategy {
  ReservedKeysVerificationStrategy(
      {bool? throwException, bool? logWarning, dynamic keys})
      : super(strategyName: 'ReservedKeysVerificationStrategy', configuration: {
          if (throwException != null) 'throwException': throwException,
          if (logWarning != null) 'logWarning': logWarning,
          if (keys != null) 'keys': keys,
        });
}

class EdgeLabelVerificationStrategy extends TraversalStrategy {
  EdgeLabelVerificationStrategy({bool? throwException, bool? logWarning})
      : super(strategyName: 'EdgeLabelVerificationStrategy', configuration: {
          if (throwException != null) 'throwException': throwException,
          if (logWarning != null) 'logWarning': logWarning,
        });
}

class MatchAlgorithmStrategy extends TraversalStrategy {
  MatchAlgorithmStrategy({String? matchAlgorithm})
      : super(strategyName: 'MatchAlgorithmStrategy', configuration: {
          if (matchAlgorithm != null) 'matchAlgorithm': matchAlgorithm,
        });
}

class HaltedTraverserStrategy extends TraversalStrategy {
  HaltedTraverserStrategy({String? haltedTraverserFactory})
      : super(strategyName: 'HaltedTraverserStrategy', configuration: {
          if (haltedTraverserFactory != null)
            'haltedTraverserFactory': haltedTraverserFactory,
        });
}
